src/main/java/org/elasticsearch/script/mustache/InternalMustacheFactory.java

```
@@ -0,0 +1,218 @@
+                return null;
+            }
+        }
+
+        @Override
+        public boolean containsKey(Object key) {
+            return get(key) != null;
+        }
+
+        @Override
+        public Set<Entry<Object, Object>> entrySet() {
+            int length = Array.getLength(array);
+            Map<Object, Object> map = new HashMap<>(length);
+            for (int i = 0; i < length; i++) {
+                map.put(i, Array.get(array, i));
+            }
+            return map.entrySet();
+        }
+
+        /**
+         * Returns an iterator over a set of elements of type T.
+         *
+         * @return an Iterator.
+         */
+        @Override
+        public Iterator<Object> iterator() {
+            return new Iter(array);
+        }
+
+        static class Iter implements Iterator<Object> {
+
+            private final Object array;
+            private final int length;
+            private int index;
+
+            public Iter(Object array) {
+                this.array = array;
+                this.length = Array.getLength(array);
+                this.index = 0;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return index < length;
+            }
+
+            @Override
+            public Object next() {
+                return Array.get(array, index++);
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException("array iterator does not support removing elements");
+            }
+        }
+    }
+
+    static class CollectionMap extends AbstractMap<Object, Object> implements Iterable<Object> {
+
+        private final Collection col;
+
+        public CollectionMap(Collection col) {
+            this.col = col;
+        }
+
+        @Override
+        public Object get(Object key) {
+            if (key instanceof Number) {
+                return Iterables.get(col, ((Number) key).intValue());
```

so I think here failing fast has a different meaning. In the sense that it seems to be that the default behaviour of `mustache` is to ignore unrecognized keys and then return an empty string. I believe we should be consistent here and not fail fast, but rather "fallback fast" :)... fallback on the default behaviour of `mustache`. This is what I do right now (after the update)... if it's non-deterministic, I fall back on the default handling of mustache.